### DISCUSSION TITLE
Error in Docs Installation

### DISCUSSION THREAD

**Discussion by michallawson** [HCL-TECH-SOFTWARE/connections-automation#50] [Q&A] [answered]
When installing Docs I receive the following Ansible error :

TASK [docs : Run the Docs Conversion installer] ********************************
fatal: [cnx07.campbellreith.com]: FAILED! => {"changed": false, "module_stderr": "Shared connection to cnx07.campbellreith.com closed.\r\n", "module_stdout": "/bin/sh: line 1: 19258 Killed                  /usr/bin/python /home/sysadmin/.ansible/tmp/ansible-tmp-1611478699.7158968-136666-214926694535902/AnsiballZ_command.py\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 137}

The log file conversion_install.log has the message:

2021-01-24 08:58:20,056 INFO server scope is Cluster
2021-01-24 08:58:20,082 INFO -->IM:Prepare installation
2021-01-24 08:58:20,083 INFO Validating installation prerequisite...
2021-01-24 08:58:20,083 INFO Start preparing conversion service installation
2021-01-24 08:58:20,083 INFO Start to show license...
2021-01-24 09:00:41,913 INFO Exception: [Errno 32] Broken pipe
2021-01-24 09:00:41,950 INFO -->IM:FAILED

Will be grateful for you assistance in resolving the problem.

The system is set up as the quickstart guide, although DB2 is is running on a separate server and cnx07.campbellreith.com is the Dmgr and WAS server.

Many thanks.

Michael Lawson

**Comment by dejo1307** [comment]
Hi Michael,

Adding @sabrina-yee  as she is more SME on the topc of Docs.

Btw. can you just try to rerun it? I did see that 'broken pipe' error before, and that mostly means that network connection broke while task was executing (or Ansible decided for whatever reason to break it, can be a hickup anywhere). Sometimes it makes sense to also go and remove the temporary folder and then rerun it again (in your case it is /home/sysadmin/.ansible on the server).

Btw. if simply rerunning it doesn't work just update it here, I will ask @sabrina-yee tomorrow to check it out during our working hours.

**Reply by michallawson** [reply]
Thank you.
Ran the installation again and also ran after deleting the tmp folder as suggested.
Both attempts failed as previously.

**Comment by sabrina-yee** [comment]
Hi @michallawson, 

Trying to isolate the issue a bit, if you go to the node where you want to install Conversion and run the installer directly (without Ansible):
```
cd /opt/HCL/InstallBinaries/Docs/DocsConversion/installer
sudo ./install.sh -wasadminID {{ was_username }} -wasadminPW {{ was_password }} -mapWebserver true -acceptLicense -silentlyInstall
```
Do you get the same error?
Also, could you confirm `/opt/HCL` have enough disk space?

**Reply by michallawson** [reply]
Thank you and have run the installation directly from the node and think we may have cracked the problem.

My password has '&' in it and when running the command line from the node only the text to the left of the & was recognised and installation failed. 

Adding quotes to the password resolved the problem and the command executed correctly and installation completed successfully.  

Very strange as it has not caused a problem with connections or customizer install.

Will uninstall and add quotes in the docs environment file and then rerun ansible to see if it resolves the issue.

Regards

Michael

**Reply by michallawson** [reply]
Installation still did not work using ansible.

One thing I have noticed is that when ansible starts running the Docs Conversion installation the amount of RAM consumed on the WAS server progressively increases until all RAM and Swap space has been used up, at which point installation fails.

**Reply by sabrina-yee** [reply]
I will try to replicate this problem internally.  It's strange because the ansible script essentially just run the same Docs installer command.  So am I understanding correctly that using ansible, it passed the `[Errno 32] Broken pipe` error but eventually failed due to out of memory?  Could you please provide the conversion_install.log?

**Reply by dejo1307** [reply]
I was seeing that broken pipe so far in two scenarios - one was when whatever network issue happens, and another was when actually disk space issues happen. For SSH, Ansible just keeps the connection, and I would also check if there is enough CPU power and if maybe kernel OOM killer kills the session once memory gets consumed. I am no expert in WAS, but I did see in multiple instances that installation itself uses like 2x RAM of what is actually needed to run the components once the installation is done.

**Reply by sabrina-yee** [reply]
What I observed is that the steps to download and unzip the Docs installers take over a gig of RAM and is not being released.  Although my all in one deployment didn't run out of memory, it was cutting thin.  If increasing RAM is not possible, as a workaround I'd suggest to comment out the download and extract then do it manually to at least see if my theory checks out.

https://github.com/HCL-TECH-SOFTWARE/connections-automation/blob/main/roles/hcl/docs/tasks/main.yml#L5-L9

The default location to store the extracted installer is `/opt/HCL/InstallBinaries/Docs`
Then extract the Conversion installer from `/opt/HCL/InstallBinaries/Docs/IBMConnectionsDocs-repo/native/DocsConversion_2.0.1.zip` to `/opt/HCL/InstallBinaries/Docs/DocsConversion`

Then run the install again to see if Conversion can be installed.

**Reply by dejo1307** [reply]
Actually, we are using unarchive module for this. It is downloading and unpacking the file 'on the fly', and considering that Docs alone is ~800MB, and that it is both downloading **and** unpacking, I can imagine it is using that amount of memory before it finishes.

**Reply by michallawson** [reply]
I am setting up a test environment to make sure the installation works, so the Dmgr and WAS are located on a single server. This has 32GB of RAM and over 20GB free space on the HDD. From what I have managed to find online I suspect the issue is related to the way Ansible runs the script and is consuming all of the available RAM.

**Reply by dejo1307** [reply]
@michallawson do you mind sharing how much RAM you had on that machine where you saw the issues initially, just to try to reproduce that as well, and see if something can be tuned on automation side?

**Comment by michallawson** [comment]
Dejan,

The server has 32GB of Ram and about 20GB of spare capacity on the HDD.

It has coped with the installation of Connections and Customizer so far 
without any problems.

I have noticed the shell module is used to run install.sh instead of the 
command module. Could this have anything to do with it?

This is the first time I have used Ansible, so no expert.

Regards

Michael

If you have received this e-mail in error please immediately notify the sender by email and delete it and any attachments from your system. This email has been sent from CampbellReith, which is the trading name of Campbell Reith Hill LLP, a limited liability partnership registered in England and Wales. Registered number, OC300082. Registered address: 15 Bermondsey Square, London, SE1 3UN. No employee or agent is authorised to conclude any binding agreement(s) on behalf of Campbell Reith Hill LLP with any other party by email unless it is an attachment on headed paper. Opinions, conclusions and other information in this email and any attachments which do not relate to the official business of Campbell Reith Hill LLP are neither given or endorsed by it. Please note that email traffic and content may be monitored.  

As this e-mail has been transmitted over a public network the accuracy, completeness and virus status of the transmitted information is not secure and cannot be guaranteed. If verification is required please telephone the sender of the email. 

This message has been scanned for malware by Websense. www.websense.com

**Comment by michallawson** [comment]
Sabrina,

Thank you.

How much ram did you have?



If you have received this e-mail in error please immediately notify the sender by email and delete it and any attachments from your system. This email has been sent from CampbellReith, which is the trading name of Campbell Reith Hill LLP, a limited liability partnership registered in England and Wales. Registered number, OC300082. Registered address: 15 Bermondsey Square, London, SE1 3UN. No employee or agent is authorised to conclude any binding agreement(s) on behalf of Campbell Reith Hill LLP with any other party by email unless it is an attachment on headed paper. Opinions, conclusions and other information in this email and any attachments which do not relate to the official business of Campbell Reith Hill LLP are neither given or endorsed by it. Please note that email traffic and content may be monitored.  

As this e-mail has been transmitted over a public network the accuracy, completeness and virus status of the transmitted information is not secure and cannot be guaranteed. If verification is required please telephone the sender of the email. 

This message has been scanned for malware by Websense. www.websense.com

**Reply by sabrina-yee** [reply]
Mine had only 16GB of RAM.

**Reply by michallawson** [reply]
Thanks

**Comment by michallawson** [comment, marked as answer]
Seems as though it was the wasadmin password which caused the issue. I reset the password and removed the "&" and the scripts are now running correctly. Not sure why that is as the password with the "&" has worked with every part of the installation up until now.

Many thanks for your help.

**Reply by dejo1307** [reply]
Thanks a lot Michael for letting us know. My guess here would be that it is a combination of & in the password and us using shell module. Shell is used specifically when you want to pass in a raw way to Ansible stuff like in this case variables, and I wouldn't be surprised if there is a bug in Ansible somewhere where it was using & the way bash would interpret it, spinning up processes in memory to the point where there was no enough memory anymore. 

I will take a look into this and try to sanitise it on our side a bit and try to reproduce then.

Cheers,
Dejan
